server/server.go

```diff
@@ -190,22 +167,56 @@ func (sc *ServerConfig) LoadFromFile(file string) error {
     sc.HistoryForwardThreshold = ysc.History.ForwardThreshold
     sc.AlertsForwardInterval = ysc.Alerts.ForwardInterval
     
-    clientCertX509, _ := x509.ParseCertificate(clientCertificate.Certificate[0])
-    serverCertX509, _ := x509.ParseCertificate(serverCertificate.Certificate[0])
-    clientCN := clientCertX509.Subject.CommonName
-    serverCN := serverCertX509.Subject.CommonName
-    
-    if len(clientCN) == 0 {
-        return errors.New("The common name in the certificate is empty. The node ID must not be empty")
-    }
-     
-    if clientCN != serverCN {
-        return errors.New(fmt.Sprintf("Server and client certificates have differing common names(%s and %s). This is the string used to uniquely identify the node.", serverCN, clientCN))
+    if (YAMLTLSFiles{}) != ysc.TLS  {
+        rootCAs := x509.NewCertPool()
+        
+        if !rootCAs.AppendCertsFromPEM([]byte(ysc.TLS.RootCA)) {
+            return errors.New("Could not append root CA to chain")
+        }
+        
+        clientCertificate, _ := tls.X509KeyPair([]byte(ysc.TLS.ClientCertificate), []byte(ysc.TLS.ClientKey))
+        serverCertificate, _ := tls.X509KeyPair([]byte(ysc.TLS.ServerCertificate), []byte(ysc.TLS.ServerKey))
+        clientTLSConfig := &tls.Config{
+            Certificates: []tls.Certificate{ clientCertificate },
+            RootCAs: rootCAs,
+            GetClientCertificate: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+                return &clientCertificate, nil
+            },
+        }
+        serverTLSConfig := &tls.Config{
+            Certificates: []tls.Certificate{ serverCertificate },
+            ClientCAs: rootCAs,
+        }
+        
+        sc.ServerTLS = serverTLSConfig
+        
+        
+        clientCertX509, _ := x509.ParseCertificate(clientCertificate.Certificate[0])
+        serverCertX509, _ := x509.ParseCertificate(serverCertificate.Certificate[0])
+        clientCN := clientCertX509.Subject.CommonName
+        serverCN := serverCertX509.Subject.CommonName
+        
+        if len(clientCN) == 0 {
+            return errors.New("The common name in the certificate is empty. The node ID must not be empty")
+        }
+         
+        if clientCN != serverCN {
+            return errors.New(fmt.Sprintf("Server and client certificates have differing common names(%s and %s). This is the string used to uniquely identify the node.", serverCN, clientCN))
+        }
+        if ysc.NodeID == ""  {
+                sc.NodeID = clientCN
+            } else {
+                sc.NodeID = ysc.NodeID
+            }
+            sc.Hub = NewHub(sc.NodeID, NewSyncController(uint(ysc.MaxSyncSessions), nil, ddbSync.NewPeriodicSyncScheduler(time.Millisecond * time.Duration(ysc.SyncSessionPeriod)), sc.SyncExplorationPathLimit), clientTLSConfig)
```

Comment:
I'd like you to try initializing variables only in one place. Use the conditionals to populate configuration then below this point call ```sc.Hub = NewHub(```

Reply:
That might mean you have to hoist variables up one block scope so they're visible beyond just the if or else clause.

Reply:
Cool